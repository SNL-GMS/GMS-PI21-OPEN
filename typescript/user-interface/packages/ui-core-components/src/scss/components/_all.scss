@import './charts/all';
@import './widgets/all';
@import './core-table';
@import './dialog';
@import './d3';
@import './divider';
@import './overlay';
@import './resizer';
@import './table-ag-grid-overrides.scss';
@import './table-overrides';
@import './table-cell';
@import './scroll-box';
@import './scroll-bar-override.scss';
