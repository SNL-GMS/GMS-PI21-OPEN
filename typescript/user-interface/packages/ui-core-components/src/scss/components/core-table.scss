/// Blueprint Table
.core-table__container {
  .bp4-table-header .bp4-table-column-name {
    height: 100%;
  }
  .bp4-table-header .bp4-table-header-content {
    position: absolute;
    margin-left: 24px;
    top: 0;
    display: inline-flex;
  }
  .core-table__column-header {
    display: flex;
    align-items: center;
    width: calc(100% - 25px);
    & .bp4-icon.bp4-intent-primary {
      color: var(--core-main, var(--gms-main, $core-main));
    }
    .force-right {
      margin-left: auto;
      margin-right: 0.5em;
    }
  }

  /// Tooltips
  .bp4-table-cell .bp4-popover-wrapper,
  .bp4-table-cell .bp4-popover-target {
    width: 100%;
  }
}
.core-table__column-header--default {
  font-size: 16px;
  font-weight: 500;
  letter-spacing: 1px;
}
