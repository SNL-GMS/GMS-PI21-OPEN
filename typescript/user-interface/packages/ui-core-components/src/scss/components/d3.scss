.tooltip-body {
  padding: 0.5em;
  font-size: 1em;
  border-radius: 3px;
  background-color: var(--core-prominent, var(--gms-prominent, $core-prominent));
  color: var(--core-main-inverted, var(--gms-main-inverted, $core-main-inverted));
  opacity: 0.9;
}

.no-pointer {
  pointer-events: none;
}

.chart__axis {
  path {
    stroke: var(--core-soft, var(--gms-soft, \$core-soft));
  }
  pointer-events: all;
}

.chart__axis-label {
  fill: var(--core-prominent, var(--gms-prominent, $core-prominent));
}

.chart-bar:hover {
  cursor: pointer;
}

.x-tick {
  font-size: 1em;
  font-family: 'Helvetica Neue', Helvetica, Arial, sans-serif, Icons16 !important;
  pointer-events: all;
  &--no-data {
    fill: var(--core-soft, var(--gms-soft, \$core-soft));
  }
}

.x-tick-container {
  pointer-events: all;
}
.chart-title {
  font-size: 1.5em;
  fill: var(--core-main, var(--gms-main, $core-main));
}
.chart-subtitle {
  font-size: 1.25em;
  fill: var(--core-prominent, var(--gms-prominent, $core-prominent));
  font-weight: 100;
  text-rendering: geometricPrecision;
}
