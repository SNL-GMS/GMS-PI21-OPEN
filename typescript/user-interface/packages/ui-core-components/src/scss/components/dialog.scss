$input-background: rgba(16, 22, 26, 0.3);
$dialog-background: var(
  --core-popover-background,
  var(--gms-popover-background, $core-popover-background)
);
$input-color: #f5f8fa;
$border-color: var(--core-recessed, var(--gms-recessed, $core-recessed));
$gap-between-rows: 0.75em;

@mixin genericInputStyles {
  font-size: 1em;
  background-color: $input-background;
  border-radius: 4px;
  color: $input-color;
  padding: 0.5em;
  border: 1px solid $border-color;
}

.dialog {
  font-size: 0.875rem; // increase or decrease this to scale the dialog
  & > * + * {
    margin-top: $gap-between-rows;
  }
  label {
    width: 100%;
    display: flex;
    flex-wrap: wrap;
    align-items: center;
    input {
      min-width: 50%;
      flex-grow: 1;
      margin: $gap-between-rows;
    }
  }
  input {
    @include genericInputStyles;
  }
  hr {
    margin-top: $gap-between-rows;
    margin-bottom: $gap-between-rows;
  }
}

.bp4-dialog.dialog_parent.dialog_parent--wide {
  width: auto;
}
.dialog__container {
  padding: 1em 1em 0.25em 1em;
}

.dialog__title {
  font-size: 1.5em;
}

.dialog__entry {
  padding: 0.5em 0;
  &:hover,
  &.selected:hover {
    cursor: pointer;
    background-color: rgba(lighten($core-table-selection, 10%), 0.9);
  }
}

.dialog__controls {
  display: flex;
  flex-wrap: nowrap;
  justify-content: space-between;
  margin-top: $gap-between-rows;
  // Target any direct children
  // & > * {
  //   // Make them at least half the width of the container, allowing for a gap
  //   min-width: calc(30% - #{$gap-between-rows});
  //   flex-grow: 1;
  // }
  // Target any direct children that have a sibling
  & > * + * {
    margin-left: $gap-between-rows; // Put a gap between them
  }
}
.dialog-actions {
  display: flex;
  margin-right: 24px;
}

.dialog-actions > button {
  margin-right: 8px;
}

// If this is a Save/Open dialog, then style the available entries
// accordingly
.dialog--save-open .dialog__entries {
  list-style: none;
  margin-top: $gap-between-rows;
  margin-bottom: $gap-between-rows;
  padding: 0.5em;
  @include genericInputStyles;
}
.dialog--save-open label input {
  width: 100%;
  margin-left: 0;
  margin-right: 0;
}

.dialog--save-open .selected {
  background-color: $core-table-selection;
}

.dialog__entry--faded {
  display: inline-block;
  color: darken($input-color, 20%);
  padding-left: 0.5em;
}
