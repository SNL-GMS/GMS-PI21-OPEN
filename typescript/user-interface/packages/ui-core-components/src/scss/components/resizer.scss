$resizer-thickness: 0.25rem;

.resizer {
  position: relative;
  transition: height linear 50ms;
  min-height: 100%;
  &__handle {
    position: absolute;
    display: block;
    margin: 0;
    border-radius: 0 0 $rounded-rectangle-radius $rounded-rectangle-radius;
    background-color: transparent;
    &--height {
      width: 100%;
      height: $resizer-thickness;
      left: 0;
      right: 0;
    }
    &--width {
      width: $resizer-thickness;
      height: 100%;
      top: 0;
      bottom: 0;
    }
    &--bottom {
      bottom: 0;
    }
    &--hint {
      background-color: var(--core-scrollbar, var(--gms-scrollbar, $core-scrollbar));
    }
    &--highlighted {
      background-color: $core-input-highlight;
    }
  }
  &__sensor {
    position: absolute;
    width: 100%;
    height: $resizer-thickness;
    padding: 0.5rem;
    background-color: transparent;
    cursor: row-resize;
    &--bottom {
      bottom: 0;
      transform: translateY(50%);
    }
  }
}
