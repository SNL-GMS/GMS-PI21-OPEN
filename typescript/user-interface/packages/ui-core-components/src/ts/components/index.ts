export * from './center-icon';
export * from './charts';
export * from './collapse';
export * from './dialog';
export * from './divider';
export * from './non-ideal-state';
export * from './overlay';
export * from './resizer';
export * from './scroll-bar-override';
export * from './scroll-box';
export * from './table';
export * from './ui-widgets';
