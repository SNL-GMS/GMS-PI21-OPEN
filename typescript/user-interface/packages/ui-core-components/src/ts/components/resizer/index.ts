// !FIX ESLINT CIRCULAR DEPENDENCY
// eslint-disable-next-line import/no-cycle
export { ResizeContainer } from './resize-container';
export { ResizeContext } from './resize-context';
export { BottomResizeHandle } from './resize-handle';
export { Resizer } from './resizer';
export * from './types';
