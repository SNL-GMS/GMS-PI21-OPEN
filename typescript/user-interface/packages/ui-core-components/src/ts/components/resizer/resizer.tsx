/* eslint-disable react/prop-types */
/* eslint-disable react/destructuring-assignment */
import type { Point } from '@gms/ui-util';
import { classList } from '@gms/ui-util';
import * as React from 'react';

import { ResizeContext } from './resize-context';
import { BottomResizeHandle } from './resize-handle';
import type { ResizeManager, ResizerProps } from './types';

const DEFAULT_CLIENT_HEIGHT_PX = 360;

/**
 * Custom hook that encapsulates the logic and exposes functions to manage resizing.
 *
 * @param onResize callback that is called and passed the current size for each rendered size during resizing.
 * @param onResizeEnd callback that is called and passed the final size after resizing finishes.
 */
const useResizeManager = (
  resizeRef: React.MutableRefObject<HTMLDivElement>,
  onResize: (h: number) => void,
  onResizeEnd: (h: number) => void
): ResizeManager => {
  const context = React.useContext(ResizeContext);

  /**
   * isResizing is a flag used to indicate if the resizer is being resized.
   */
  const [isResizing, setResizing] = React.useState(false);

  /**
   * initialHeight is the height of the resizer when the user starts resizing it.
   * Combined with the amount the mouse has moved, this determines the user's
   * intended height.
   */
  const initialHeightRef = React.useRef(context.height);

  // Getters
  const getIsResizing = () => isResizing;

  /**
   * the actual height of the resizer in the DOM
   */
  const getCurrentHeight = React.useCallback(
    () =>
      Math.min(resizeRef.current?.clientHeight, context.containerHeight) ||
      resizeRef.current?.clientHeight ||
      context.height,
    [context.containerHeight, context.height, resizeRef]
  );

  /**
   * The height the resizer says we should currently use
   */
  const getHeight = () =>
    context.height && context.containerHeight
      ? Math.min(context.height, context.containerHeight) || context.height
      : DEFAULT_CLIENT_HEIGHT_PX;

  // Init
  React.useLayoutEffect(() => {
    const h = getCurrentHeight();
    context.setHeight(h);
    initialHeightRef.current = h;
  }, [context, getCurrentHeight]);

  // Setters
  const setIsResizing = (isIt: boolean) => {
    context.setIsResizing(isIt); // updates the context
    setResizing(isIt); // our locally managed isResizing state
  };

  // Event handlers
  /**
   * Handle mouse move while dragging
   */
  const handleMouseMove = (startPoint: Point) => (e2: MouseEvent) => {
    const mouseMoveFromStart = e2.pageY - (startPoint?.y ?? e2.pageY);
    const initialHeight = initialHeightRef.current;
    const newHeight = Math.min(initialHeight + mouseMoveFromStart, context.containerHeight);
    context.setHeight(newHeight);
    if (onResize) {
      onResize(newHeight);
    }
  };

  /**
   * Handle the end of resizing. Called by onMouseUp.
   */
  const handleResizeEnd = () => {
    setResizing(false);
    context.setIsResizing(false);
    context.setHeight(getCurrentHeight());
    onResizeEnd(context.height);
  };

  /**
   * handle the start of a resize event. Called by onMouseDown.
   */
  const handleResizeStart = () => {
    setIsResizing(true);
    initialHeightRef.current = getCurrentHeight();
  };

  return {
    getIsResizing,
    getHeight,
    handleMouseMove,
    handleResizeStart,
    handleResizeEnd
  };
};

// TODO: Create options for different orientations. Currently only supports BOTTOM, by default.
/**
 * A resizer component that creates a drag handle and sensor, and updates its size when that
 * handle is dragged. Synchs sizes with other resizer components within the same ResizeContainer.
 */
// eslint-disable-next-line react/function-component-definition
export const Resizer: React.FunctionComponent<React.PropsWithChildren<ResizerProps>> = props => {
  const context = React.useContext(ResizeContext);
  // !FIX ESLINT DO NOT USE REACT HOOK HOOKS IN CONDITIONAL
  // eslint-disable-next-line react-hooks/rules-of-hooks
  const resizeRef = props.forwardRef ? props.forwardRef : React.useRef<HTMLDivElement>(null);
  const resizeManager = useResizeManager(
    resizeRef,
    (h: number) => props.onResize(h),
    (h: number) => props.onResizeEnd(h)
  );
  return (
    <div
      className={classList(
        {
          resizer: true,
          'resizer--resizing': resizeManager.getIsResizing()
        },
        props.className ?? ''
      )}
      data-cy={props.dataCy}
      ref={resizeRef}
      style={{
        height: Math.max(props?.minHeightPx ?? 0, resizeManager.getHeight()),
        maxHeight: props.maxHeightPx ?? context.containerHeight ?? '100vh'
      }}
    >
      {props.children}
      <BottomResizeHandle
        handleMouseMove={e => resizeManager.handleMouseMove(e)}
        onResizeEnd={() => resizeManager.handleResizeEnd()}
        onResizeStart={() => resizeManager.handleResizeStart()}
      />
    </div>
  );
};
