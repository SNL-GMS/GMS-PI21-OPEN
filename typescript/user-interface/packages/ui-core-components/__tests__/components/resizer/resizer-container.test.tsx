import { render } from '@testing-library/react';
import React from 'react';

import { ResizeContainer } from '../../../src/ts/components/resizer';

// set up window alert and open so we don't see errors
(window as any).alert = jest.fn();
(window as any).open = jest.fn();

describe('ResizeContainer', () => {
  const { container } = render(
    <ResizeContainer>
      <div />
    </ResizeContainer>
  );
  it('is exported', () => {
    expect(ResizeContainer).toBeDefined();
  });
  // TODO Make new tests this tests nothing
  it.skip('will match its snapshot', () => {
    expect(container).toMatchSnapshot();
  });
});
