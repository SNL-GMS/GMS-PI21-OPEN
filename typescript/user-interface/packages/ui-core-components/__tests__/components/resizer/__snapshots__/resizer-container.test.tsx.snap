// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResizeContainer will match its snapshot 1`] = `<div />`;
